### DISCUSSION TITLE
Add attributes to user during authorization

### DISCUSSION THREAD

**Discussion by danjamesmedia** [nextauthjs/next-auth#764] [Help] [answered]
Whenever I try to add more attributes to the user object in a custom authorization call, it seems I can only set the name, email and image. I would like to add an Id field based on a value retrieved in the call made to my external API during the call.

I see the other way to set this is using the session callback, but this would involve an additional API call to get the info.

Is there any way that more attributes can be added to the user object/session during the authorization process?

**Comment by dimbslmh** [comment, marked as answer]
I was having the same issue, but it has been documented here https://next-auth.js.org/configuration/callbacks#session-callback:

> When using JSON Web Tokens for sessions, the JWT payload is provided instead.

> ### Tip
> When using JSON Web Tokens the jwt() callback is invoked before the session() callback, so anything you add to the JSON Web Token will be immediately available in the session callback.

> ### Tip
> To better represent its value, when using a JWT session, the second parameter should be called token (This is the same thing you return from the jwt() callback). ...

In the jwt callback you can add attributes from the user object to the token object that you want to have during the session callback as the second parameter.

**Reply by Fziliotti** [reply]
I'm still having trouble to return new attributes.

I use Credentials Provider, and I tested it by adding the session() and jwt () callbacks but I couldn't get it to work.

**Reply by Fziliotti** [reply]
The response from ''http://.../api/auth/session" is always an object like:

`{"user":{"name":"Joseph","email":"test@test.com","image":null},"expires":"2021-04-07T01:08:19.405Z"}`

**Reply by balazsorban44** [reply]
Make sure that you use a conditional branch in your `jwt`. Read the docs carefully: https://next-auth.js.org/configuration/callbacks#jwt-callback. The `jwt` callback will return the user data only the first time it is invoked. You should check if user exists, and only then return the data. Otherwise, all subsequent `jwt` calls will just cancel out your additions.

**Reply by Fziliotti** [reply]
@balazsorban44  Thanks for the reply! I read the documentation, but I cant resolve the problem.

I have an authorize() function that makes a fetch and returns an object with some attributes (like an clientId).

And I configurate the jwt and session callbacks with the conditional branch, but the response still not returning the other attributes, Its like a filter that just accepts name, email and image.

I just want to pass a clientId to the session object...

**Reply by Fziliotti** [reply]
I am receiving the warning from jwt_auto_generated_signing_key, could this be influencing? I think not

**Reply by balazsorban44** [reply]
at this point, I ask you to open a new issue with a reproduction. It is something that is misconfigured on your end, I think.

the warning you are getting is irrelevant for your problem. Although you can fix it https://next-auth.js.org/warnings#jwt_auto_generated_signing_key

**Reply by jackzbazuka** [reply]
What if I'm using just a simple reddit sign in? I can only access the name, image and email. I want add the user's reddit id to my session.user, so I can access it on client side with session.user.uid

**Reply by balazsorban44** [reply]
Then I recommend reading the docs:

https://next-auth.js.org/configuration/callbacks#jwt-callback

https://next-auth.js.org/configuration/callbacks#session-callback

**Reply by raulmarindev** [reply]
Would it make sense to add a more specific example to the Callbacks section in the docs? I'm thinking of an example that shows all the changes needed in order to have custom attributes in a user. The reason being it's a very common scenario and users would benefit from it.  Something similar to what was provided by @MateuszWawrzynski in his answer could be enough.

**Reply by ogooluwanick** [reply]
this is awesome, thanks .Although I'm not sure I fully understand the docs, I just added the missing field I needed to the jwt and session callback and I still returned the field in CredentialsProvider{authorize}

**Reply by derekgygax** [reply]
Yes, I more specific example would be extremely helpful in the docs.

**Comment by MateuszWawrzynski** [comment]
I had the same problem, but resolved it.
That thread helped me so much.
Thanks for answer @dimbslmh 👍 

Explanation here:
https://stackoverflow.com/a/64595973/14536846

**Reply by nemanjam** [reply]
But now `session` is defined even when user is not logged in. So if I want to check if user is logged in I need to check `session.user`. It didn't work that way originally, is this good solution at all?

```javascript
console.log(session) // non logged in user
Object { expires: "2021-12-19T20:55:44.430Z" }
```

**Reply by baptisteArno** [reply]
I'm wondering why it isn't the default behavior

```ts
    jwt: async ({ token, user }) => {
      user && (token.user = user)
      return token
    },
    session: async ({ session, token }) => {
      if (token.user) session.user = token.user as any
      return session
    },
```

at least for `CredentialsProvider`, should I create a PR for this @balazsorban44 ?

**Reply by balazsorban44** [reply]
No. This is on purpose. Check the documentation.

**Reply by baptisteArno** [reply]
```
NextAuth.js does not limit how much data you can store in a JSON Web Token, however a ~4096 byte limit per cookie is commonly imposed by browsers.
```

This? 👆

**Reply by balazsorban44** [reply]
https://next-auth.js.org/configuration/callbacks#session-callback

> By default, only a subset of the token is returned for increased security. If you want to make something available you added to the token through the jwt() callback, you have to explicitly forward it here to make it available to the client.

**Reply by nemanjam** [reply]
I had to do this so that I have falsy session for non-logged in user.

```typescript
async session({ session, token }) {
  const _session = token.user ? { ...session, user: token.user } : undefined; // set undefined explicitly 
  return _session as Session;
},

```

**Reply by balazsorban44** [reply]
If a user is not logged in, there is no session, so that would be unnecessary.

**Reply by phongvuhyuet** [reply]
I have another issue, because my user object is quite big so the token will be very heavy. I'd prefer not passing the user object to the token, is there any way else? 

<img width="1080" alt="image" src="https://user-images.githubusercontent.com/70512647/230004150-6fd3dfd9-3c5c-4947-a5bb-674d5dc04917.png">

**Comment by nolawnchairs** [comment]
This is way more difficult than it needs to be.

**Comment by fFernandoh** [comment]
I find it a bit confusing what is what in nextAuth, but what I did and it worked was:

```
declare module "next-auth" {
  interface Session extends DefaultSession {
    user: {
      id: string;
      role: UserRole;
      emailVerified: Date;
      client: Client & {
        favorites: FictitiousProfileFavorite[];
      };
      yourData: YourType;
    } & DefaultSession["user"];
  }
}
```

in the CredentialsProviders: 

```
 async authorize(data) {
        const email: string = data.email as string;

        const user = await db.user.findUnique({
          where: { email: email },
          include: {
            client: {
              include: {
                favorites: true,
              },
            },
          },
        });
        return user;
        
```

**Reply by BugKiller-tech** [reply]
Cool, This is what I was looking for.
Thank you
